opencog/attention/experimental/ForgettingAgent.cc

```diff
@@ -0,0 +1,146 @@
+ */
+
+#include <algorithm>
+#include <sstream>
+
+#define DEPRECATED_ATOMSPACE_CALLS
+#include <opencog/atomspace/AtomSpace.h>
+
+#include <opencog/cogserver/server/Agent.h>
+#include <opencog/cogserver/server/CogServer.h>
+#include <opencog/cogserver/server/Factory.h>
+#include <opencog/attention/atom_types.h>
+#include <opencog/util/Config.h>
+#include "ForgettingAgent.h"
+
+using namespace opencog;
+
+ForgettingAgent::ForgettingAgent(CogServer& cs) :
+    Agent(cs)
+{
+    std::string defaultForgetThreshold;
+    std::ostringstream buf;
+
+    // No limit to lti of removed atoms
+    // Convert MAXLTI to a string for storing in the configuration
+    buf << AttentionValue::MAXLTI;
+    defaultForgetThreshold = buf.str();
+    config().set("ECAN_FORGET_THRESHOLD",defaultForgetThreshold);
+
+    forgetThreshold = (AttentionValue::lti_t)
+                      (config().get_int("ECAN_FORGET_THRESHOLD"));
+
+    //Todo: Make configurable
+    maxSize = config().get_int("ECAN_ATOMSPACE_MAXSIZE");
+    accDivSize = config().get_int("ECAN_ATOMSPACE_ACCEPTABLE_SIZE_SPREAD");
+
+    // Provide a logger, but disable it initially
+    log = NULL;
+    setLogger(new opencog::Logger("ForgettingAgent.log", Logger::WARN, true));
+}
+
+ForgettingAgent::~ForgettingAgent()
+{
+    if (log) delete log;
+}
+
+Logger* ForgettingAgent::getLogger()
+{
+    return log;
+}
+
+void ForgettingAgent::setLogger(Logger* _log)
+{
+    if (log) delete log;
+    log = _log;
+}
+
+void ForgettingAgent::run()
+{
+    log->fine("=========== ForgettingAgent::run =======");
+    a = &_cogserver.getAtomSpace();
+    forget();
+}
+
+void ForgettingAgent::forget()
+{
+    HandleSeq atomsVector;
+    std::back_insert_iterator<HandleSeq> output2(atomsVector);
+    int count = 0;
+    int removalAmount;
+    bool recursive;
+
+    a->get_handles_by_type(output2, ATOM, true);
+
+    int asize = atomsVector.size();
+    if (asize < (maxSize + accDivSize)) {
+        return;
+    }
+
+    fprintf(stdout,"Forgetting Stuff, Atomspace Size: %d \n",asize);
+    // Sort atoms by lti, remove the lowest unless vlti is NONDISPOSABLE
+    std::sort(atomsVector.begin(), atomsVector.end(), ForgettingLTIThenTVAscendingSort(a));
+
+    removalAmount = asize - (maxSize - accDivSize);
+    log->info("ForgettingAgent::forget - will attempt to remove %d atoms", removalAmount);
+
+    for (unsigned int i = 0; i < atomsVector.size(); i++)
+    {
+        if (atomsVector[i]->getAttentionValue()->getLTI() <= forgetThreshold
+                && count < removalAmount)
+        {
+            if (atomsVector[i]->getAttentionValue()->getVLTI() == AttentionValue::DISPOSABLE )
+            {
+                std::string atomName = a->atom_as_string(atomsVector[i]);
+                log->fine("Removing atom %s", atomName.c_str());
+                // TODO: do recursive remove if neighbours are not very important
+                IncomingSet iset = atomsVector[i]->getIncomingSet(a);
```

Comment:
```
const IncomingSet&
```
The ampersand avoids a copy of the entire contents of the vector,  without the ampersand, the vector is copied, and you really don't need a copy here.